File: test/benchmark/r7rs/python/mbrot.py
#!/usr/bin/env python3
# R7RS Benchmark: mbrot (Python)
# Mandelbrot set generation - 75x75 grid, result at (0,0) = 5
import time


def count(r, i, step, x, y):
    max_count = 64
    radius2 = 16.0
    cr = r + x * step
    ci = i + y * step
    zr = cr
    zi = ci
    c = 0
    while c < max_count:
        zr2 = zr * zr
        zi2 = zi * zi
        if zr2 + zi2 > radius2:
            return c
        new_zr = zr2 - zi2 + cr
        zi = 2.0 * zr * zi + ci
        zr = new_zr
        c += 1
    return max_count


def mbrot(matrix, r, i, step, n):
    for y in range(n - 1, -1, -1):
        for x in range(n - 1, -1, -1):
            matrix[x][y] = count(r, i, step, x, y)


def test(n):
    matrix = [[0] * n for _ in range(n)]
    mbrot(matrix, -1.0, -0.5, 0.005, n)
    return matrix[0][0]


def main():
    t0 = time.perf_counter_ns()
    result = test(75)
    t1 = time.perf_counter_ns()

    if result == 5:
        print("mbrot: PASS")
    else:
        print(f"mbrot: FAIL result={result}")
    print(f"__TIMING__:{(t1 - t0) / 1e6:.3f}")


main()

File: test/benchmark/r7rs/c2mir/mbrot.c
/* Native C2MIR port of r7rs/mbrot2.ls. */
extern int printf(const char *, ...);

static int count(double r, double i, double step, int x, int y) {
    double cr = r + x * step;
    double ci = i + y * step;
    double zr = cr;
    double zi = ci;
    int c;
    for (c = 0; c < 64; c++) {
        double zr2 = zr * zr;
        double zi2 = zi * zi;
        double new_zr;
        if (zr2 + zi2 > 16.0) return c;
        new_zr = zr2 - zi2 + cr;
        zi = 2.0 * zr * zi + ci;
        zr = new_zr;
    }
    return 64;
}

int main(void) {
    int matrix[75][75];
    int x;
    int y;
    int result;
    for (y = 74; y >= 0; y--) for (x = 74; x >= 0; x--) matrix[x][y] = count(-1.0, -0.5, 0.005, x, y);
    result = matrix[0][0];
    printf(result == 5 ? "mbrot: PASS\n" : "mbrot: FAIL result=%d\n", result);
    return result != 5;
}
